Add margin when comptuing baseline position for tables.

sammygill opened the pull request:
#### c0048e83eb237ba2dba8a01bfc46bf0de42412a2
<pre>
Add margin when comptuing baseline position for tables.
<a href="https://bugs.webkit.org/show_bug.cgi?id=97239">https://bugs.webkit.org/show_bug.cgi?id=97239</a>
rdar://problem/100910532

Reviewed by Brent Fulgham.

After computing the baseline of a table, we need to take into
consideration any margins that are shifting the element. Without adding
the margin to the baseline, any items that are basline against the table
may become offset by that amount.

* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::baselinePosition const):

Canonical link: <a href="https://commits.webkit.org/255357@main">https://commits.webkit.org/255357@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e32f9b57b4b6ce2456628d7719f92096437a52ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22688 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101944 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/162170 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1385 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29781 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84599 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98111 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/891 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78680 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27832 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82810 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82414 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70873 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36205 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16416 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33962 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17545 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3713 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37832 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40227 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39732 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36671 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline.html:
```diff
@@ -0,0 +1,39 @@
+ }
+ .v-table {
+     margin-right: 50px;
+ }
```
Ditto replacing with a logical property

sammygill replied:
Changed to logical properties

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline.html:
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+ <html>
```
Nit: this file has an indentation of 1 space that seems unnecessary.

sammygill replied:
Indentation should be fixed

brentfulgham commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1508,9 +1508,9 @@ RenderTableCell* RenderTable::cellAfter(const RenderTableCell* cell) const
-        return RenderBox::baselinePosition(baselineType, firstLine, direction, linePositionMode);
-    });
+    if (auto baselinePos = firstLineBaseline())
+        return (direction == HorizontalLine ? marginTop() : marginRight()) + firstLineBaseline().value();
```
I think you want to change "firstLineBaseline().value()" to just "baselinePos.value()", since you just captured that variable in line 1511. No need to call that accessor again.

Or, you could do:

```
if (auto baselinePos = firstLineBaseline())
    return (direction == HorizontalLine ? marginTop() : marginRight()) + *baselinePos;
```

sammygill replied:
Whoops yes thank you. Changed.

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline-expected.html:
```diff
@@ -0,0 +1,43 @@
+    border: solid 1px black;
+}
+.vertical {
+  writing-mode: vertical-rl
```
Nit: inconsistent indentation and trailing semicolon

sammygill replied:
Fixed

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline-expected.html:
```diff
@@ -0,0 +1,43 @@
+  writing-mode: vertical-rl
+}
+.block {
+    display:inline-block;
```
Nit: `display: inline-block;` with a space, for consistency

sammygill replied:
Fixed

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline-expected.html:
```diff
@@ -0,0 +1,43 @@
+}
+.v-margin {
+    margin-right: 50px;
+}
```
Nit: unless you think it's important to use physical properties, you could use `margin-block-start: 50px` in `.table` and remove `.h-table` and `.v-table`.

sammygill replied:
Used logical properties

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline.html:
```diff
@@ -0,0 +1,39 @@
+     border: solid 1px black;
+ }
+ .vertical {
+     writing-mode: vertical-rl
```
Ditto semicolon

sammygill replied:
Fixed

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline.html:
```diff
@@ -0,0 +1,39 @@
+     writing-mode: vertical-rl
+}
+ .block {
+     display:inline-block;
```
Ditto space after colon

sammygill replied:
Fixed

Loirooriol commented on LayoutTests/fast/table/inline-table-margin-baseline-expected.html:
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
```
Firefox is failing this test in the vertical case, so it would be good to make it a WPT test.

sammygill replied:
Created WPT PR: https://github.com/web-platform-tests/wpt/pull/36337

Loirooriol reviewed (approved):
Informal LGTM with suggested changes

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/baseline-rules/inline-table-inline-block-baseline-expected.html:
```diff
@@ -0,0 +1,36 @@
+}
+.vertical {
+    writing-mode: vertical-rl;
+}
```
Nit: this is not used

sammygill replied:
Removed

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/baseline-rules/inline-table-inline-block-baseline-ref.html:
```diff
@@ -0,0 +1,36 @@
+}
+.vertical {
+    writing-mode: vertical-rl;
+}
```
Ditto.

sammygill replied:
Removed

sammygill committed:
Add margin when comptuing baseline position for tables.
https://bugs.webkit.org/show_bug.cgi?id=97239
rdar://problem/100910532

Reviewed by Brent Fulgham.

After computing the baseline of a table, we need to take into
consideration any margins that are shifting the element. Without adding
the margin to the baseline, any items that are basline against the table
may become offset by that amount.

* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::baselinePosition const):

Canonical link: https://commits.webkit.org/255357@main
